File: scripts/contract_smoke.py
#!/usr/bin/env python3
"""
Backend API Contract Smoke Test

Tests the core API endpoints defined in openapi/backend-api.openapi.yaml
to verify the backend is functioning correctly.

Usage:
    python contract_smoke.py --base-url http://127.0.0.1:8788 --project-slug smoke_test
    python contract_smoke.py --base-url http://127.0.0.1:8788 --api-token YOUR_TOKEN
"""
from __future__ import annotations

import argparse
import json
import urllib.error
import urllib.request


def request_json(base_url: str, method: str, path: str, payload: dict | None = None, api_token: str | None = None) -> dict:
    """Make HTTP request and return JSON response."""
    data = None
    headers = {"Content-Type": "application/json"}
    if api_token:
        headers["Authorization"] = f"Bearer {api_token}"
    if payload is not None:
        data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(f"{base_url}{path}", data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:  # noqa: S310
            raw = resp.read().decode("utf-8")
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8", errors="replace")
        raise SystemExit(f"{method} {path} failed: HTTP {exc.code} {body}") from exc
    except urllib.error.URLError as exc:
        raise SystemExit(f"{method} {path} failed: {exc.reason}") from exc
    if not raw:
        return {}
    return json.loads(raw)


def main():
    parser = argparse.ArgumentParser(description="Backend API contract smoke test")
    parser.add_argument("--base-url", default="http://127.0.0.1:8788", help="Backend API base URL")
    parser.add_argument("--project-slug", default="contract_demo", help="Project slug to use")
    parser.add_argument("--api-token", help="API bearer token (optional if KROMA_API_AUTH_DEV_BYPASS=true)")
    args = parser.parse_args()

    base_url = args.base_url.rstrip("/")
    slug = args.project_slug.strip() or "contract_demo"
    api_token = args.api_token

    print("[contract-smoke] Testing backend API contract")
    print(f"[contract-smoke] Base URL: {base_url}")
    print(f"[contract-smoke] Project slug: {slug}")
    if api_token:
        print(f"[contract-smoke] Using API token: {api_token[:8]}...")
    else:
        print("[contract-smoke] No API token provided - requires KROMA_API_AUTH_DEV_BYPASS=true")
    print()

    # Health check (no auth required)
    print("[contract-smoke] 1/8 - Health check")
    health = request_json(base_url, "GET", "/health", api_token=api_token)
    assert health.get("ok") is True, "Health check failed"
    print(f"           ✓ Backend healthy: {health.get('service')} v{health.get('version')}")
    print(f"           ✓ Routes: {health.get('route_count')}")

    # Create project
    print("[contract-smoke] 2/8 - Upsert project")
    project = request_json(base_url, "POST", "/api/projects", {"name": "Contract Demo", "slug": slug}, api_token=api_token)
    assert "project" in project, "Project creation failed"
    print(f"           ✓ Project created: {slug}")

    # Prompt templates CRUD
    print("[contract-smoke] 3/8 - Prompt templates CRUD")
    created_template = request_json(
        base_url,
        "POST",
        f"/api/projects/{slug}/prompt-templates",
        {"name": "default-shot", "template_text": "A cinematic shot of {subject}"},
        api_token=api_token,
    )
    assert "prompt_template" in created_template, "Template creation failed"
    template_id = created_template["prompt_template"]["id"]
    print(f"           ✓ Template created: {template_id}")

    templates_list = request_json(base_url, "GET", f"/api/projects/{slug}/prompt-templates", api_token=api_token)
    assert "templates" in templates_list or "prompt_templates" in templates_list, "Template list failed"

    template_detail = request_json(base_url, "GET", f"/api/projects/{slug}/prompt-templates/{template_id}", api_token=api_token)
    assert "prompt_template" in template_detail, "Template detail failed"

    updated = request_json(
        base_url,
        "PUT",
        f"/api/projects/{slug}/prompt-templates/{template_id}",
        {"template_text": "A cinematic close-up of {subject}"},
        api_token=api_token,
    )
    assert updated.get("ok") is True, "Template update failed"

    deleted = request_json(base_url, "DELETE", f"/api/projects/{slug}/prompt-templates/{template_id}", api_token=api_token)
    assert deleted.get("ok") is True, "Template deletion failed"
    print("           ✓ Templates CRUD complete")

    # Chat sessions
    print("[contract-smoke] 4/8 - Chat sessions")
    sess = request_json(base_url, "POST", f"/api/projects/{slug}/chat/sessions", {"title": "Contract session"}, api_token=api_token)
    assert "session" in sess, "Session creation failed"
    session_id = sess["session"]["id"]
    print(f"           ✓ Session created: {session_id}")

    # Add message
    print("[contract-smoke] 5/8 - Add chat message")
    msg = request_json(
        base_url,
        "POST",
        f"/api/projects/{slug}/chat/sessions/{session_id}/messages",
        {"role": "user", "content_text": "Create instruction"},
        api_token=api_token,
    )
    assert "message" in msg, "Message creation failed"
    print("           ✓ Message added")

    # Agent instructions
    print("[contract-smoke] 6/8 - Agent instructions")
    ins = request_json(
        base_url,
        "POST",
        f"/api/projects/{slug}/agent/instructions",
        {"instruction_type": "pipeline_run", "dispatch_to_agent": False, "payload_json": {"stage": "style", "candidates": 2}},
        api_token=api_token,
    )
    assert "instruction" in ins, "Instruction creation failed"
    instr_id = ins["instruction"]["id"]
    print(f"           ✓ Instruction created: {instr_id}")

    # Events
    print("[contract-smoke] 7/8 - Instruction events")
    events = request_json(base_url, "GET", f"/api/projects/{slug}/agent/instructions/{instr_id}/events", api_token=api_token)
    assert "events" in events or "instruction_events" in events, "Events fetch failed"
    print("           ✓ Events retrieved")

    # Secrets CRUD
    print("[contract-smoke] 8/8 - Project secrets")
    secret_resp = request_json(
        base_url,
        "POST",
        f"/api/projects/{slug}/secrets",
        {"provider_code": "openai", "secret_name": "api_key", "secret_value": "sk-test-contract-123456"},
        api_token=api_token,
    )
    assert secret_resp.get("ok") is True, "Secret creation failed"
    print("           ✓ Secret created")

    secrets_list = request_json(base_url, "GET", f"/api/projects/{slug}/secrets", api_token=api_token)
    assert "secrets" in secrets_list, "Secrets list failed"

    secret_delete = request_json(base_url, "DELETE", f"/api/projects/{slug}/secrets/openai/api_key", api_token=api_token)
    assert secret_delete.get("ok") is True, "Secret deletion failed"
    print("           ✓ Secrets CRUD complete")

    # Exports list
    print("[contract-smoke] Exports list")
    exports = request_json(base_url, "GET", f"/api/projects/{slug}/exports", api_token=api_token)
    assert "project_exports" in exports or "exports" in exports, "Exports response should include exports list"
    print("           ✓ Exports retrieved")

    print()
    print(f"[contract-smoke] ✓ All tests passed! project={slug}")
    print(f"[contract-smoke] Created resources: template={template_id}, session={session_id}, instruction={instr_id}")
    print()
    print("[contract-smoke] NOTE: To enable auth for local development, set:")
    print("                     KROMA_API_AUTH_DEV_BYPASS=true")


if __name__ == "__main__":
    main()

File: src-tauri/tests/contract_smoke.rs
use std::path::PathBuf;
use std::sync::Arc;

use axum::body::{to_bytes, Body};
use axum::http::{Method, Request, StatusCode};
use serde_json::{json, Value};
use tower::ServiceExt;
use uuid::Uuid;

use kroma_backend_core::api::server::build_router_with_projects_store_dev_bypass;
use kroma_backend_core::db::projects::ProjectsStore;

#[tokio::test]
async fn contract_smoke_test_full_flow() {
    let app = build_router_with_projects_store_dev_bypass(test_store());
    let slug = "contract_demo";

    let _project = send_json(
        app.clone(),
        Method::POST,
        "/api/projects",
        Body::from(json!({"name": "Contract Demo", "slug": slug}).to_string()),
        StatusCode::OK,
    )
    .await;

    let created_template = send_json(
        app.clone(),
        Method::POST,
        &format!("/api/projects/{slug}/prompt-templates"),
        Body::from(json!({"name": "default-shot", "template_text": "A cinematic shot of {subject}"}).to_string()),
        StatusCode::OK,
    )
    .await;
    let template_id = created_template["prompt_template"]["id"]
        .as_str()
        .expect("template id should exist")
        .to_string();

    let _templates_list = send_json(
        app.clone(),
        Method::GET,
        &format!("/api/projects/{slug}/prompt-templates"),
        Body::empty(),
        StatusCode::OK,
    )
    .await;

    let _template_detail = send_json(
        app.clone(),
        Method::GET,
        &format!("/api/projects/{slug}/prompt-templates/{template_id}"),
        Body::empty(),
        StatusCode::OK,
    )
    .await;

    let _updated = send_json(
        app.clone(),
        Method::PUT,
        &format!("/api/projects/{slug}/prompt-templates/{template_id}"),
        Body::from(json!({"template_text": "A cinematic close-up of {subject}"}).to_string()),
        StatusCode::OK,
    )
    .await;

    let _deleted = send_json(
        app.clone(),
        Method::DELETE,
        &format!("/api/projects/{slug}/prompt-templates/{template_id}"),
        Body::empty(),
        StatusCode::OK,
    )
    .await;

    let sess = send_json(
        app.clone(),
        Method::POST,
        &format!("/api/projects/{slug}/chat/sessions"),
        Body::from(json!({"title": "Contract session"}).to_string()),
        StatusCode::OK,
    )
    .await;
    let session_id = sess["session"]["id"]
        .as_str()
        .expect("session id should exist")
        .to_string();

    let _message = send_json(
        app.clone(),
        Method::POST,
        &format!("/api/projects/{slug}/chat/sessions/{session_id}/messages"),
        Body::from(json!({"role": "user", "content": "Create instruction"}).to_string()),
        StatusCode::OK,
    )
    .await;

    let ins = send_json(
        app.clone(),
        Method::POST,
        &format!("/api/projects/{slug}/agent/instructions"),
        Body::from(json!({"instruction_text": "Test instruction"}).to_string()),
        StatusCode::OK,
    )
    .await;
    let instr_id = ins["instruction"]["id"]
        .as_str()
        .expect("instruction id should exist")
        .to_string();

    let _events = send_json(
        app.clone(),
        Method::GET,
        &format!("/api/projects/{slug}/agent/instructions/{instr_id}/events"),
        Body::empty(),
        StatusCode::OK,
    )
    .await;

    let _secret_created = send_json(
        app.clone(),
        Method::POST,
        &format!("/api/projects/{slug}/secrets"),
        Body::from(json!({"provider_code": "openai", "secret_name": "api_key", "secret_value": "sk-test-contract-123456"}).to_string()),
        StatusCode::OK,
    )
    .await;

    let secrets_list = send_json(
        app.clone(),
        Method::GET,
        &format!("/api/projects/{slug}/secrets"),
        Body::empty(),
        StatusCode::OK,
    )
    .await;
    assert_eq!(secrets_list["count"].as_i64().expect("count should be integer"), 1);

    let _secret_deleted = send_json(
        app.clone(),
        Method::DELETE,
        &format!("/api/projects/{slug}/secrets/openai/api_key"),
        Body::empty(),
        StatusCode::OK,
    )
    .await;

    let exports = send_json(
        app,
        Method::GET,
        &format!("/api/projects/{slug}/exports"),
        Body::empty(),
        StatusCode::OK,
    )
    .await;
    assert_eq!(exports["ok"], json!(true));

    println!("[contract-smoke] ok project={slug} template={template_id} session={session_id} instruction={instr_id}");
}

async fn send_json(
    app: axum::Router,
    method: Method,
    uri: &str,
    body: Body,
    expected_status: StatusCode,
) -> Value {
    let request = Request::builder()
        .method(method)
        .uri(uri)
        .header("content-type", "application/json")
        .body(body)
        .expect("request should build");

    let response = app
        .oneshot(request)
        .await
        .expect("router should return response");
    assert_eq!(response.status(), expected_status, "Expected status {} but got {}", expected_status, response.status());
    let body = to_bytes(response.into_body(), usize::MAX)
        .await
        .expect("body should be readable");
    serde_json::from_slice(body.as_ref()).expect("response should be valid JSON")
}

fn test_store() -> Arc<ProjectsStore> {
    let suffix = Uuid::new_v4().to_string();
    let root = std::env::temp_dir().join(format!("kroma_contract_smoke_test_{suffix}"));
    let db = root.join("var/backend/app.db");
    std::fs::create_dir_all(root.as_path()).expect("temp test root must be creatable");
    let store = Arc::new(ProjectsStore::new(db, PathBuf::from(root)));
    store.initialize().expect("store should initialize");
    store
}
